python_modules/dagster/dagster_tests/general_tests/parent_asset_graph_differ_tests/test_parent_asset_graph_diff.py

```diff
@@ -0,0 +1,151 @@
+from dagster._core.workspace.context import WorkspaceRequestContext
+from dagster._core.workspace.workspace import (
+    CodeLocationEntry,
+    CodeLocationLoadStatus,
+)
+
+from .parent_deployment_asset_graphs.basic_asset_graph import defs as basic_asset_graph
+
+parent_deployment_defs_by_name = {"basic_asset_graph": basic_asset_graph}
+
+
+@pytest.fixture
+def instance():
+    with instance_for_test() as the_instance:
+        yield the_instance
+
+
+def _make_location_entry(parent_graph_name: str, instance: DagsterInstance):
+    origin = InProcessCodeLocationOrigin(
+        loadable_target_origin=LoadableTargetOrigin(
+            executable_path=sys.executable,
+            module_name=(
+                f"dagster_tests.general_tests.asset_graph_diff_tests.parent_deployment_asset_graphs.{parent_graph_name}"
+            ),
+            working_directory=os.getcwd(),
+            attribute="defs",
+        ),
+        container_image=None,
+        entry_point=None,
+        container_context=None,
+        location_name=None,
+    )
+
+    code_location = origin.create_location(instance)
+
+    return CodeLocationEntry(
+        origin=origin,
+        code_location=code_location,
+        load_error=None,
+        load_status=CodeLocationLoadStatus.LOADED,
+        display_metadata={},
+        update_timestamp=time.time(),
+    )
+
+
+def _make_workspace_context(instance: DagsterInstance, parent_graph_names):
+    return WorkspaceRequestContext(
+        instance=mock.MagicMock(),
+        workspace_snapshot={
+            name: _make_location_entry(name, instance) for name in parent_graph_names
+        },
+        process_context=mock.MagicMock(),
+        version=None,
+        source=None,
+        read_only=True,
+    )
+
+
+def _get_branch_deployment_graph_with_code_changes(
+    parent_graph_name, new_assets: Optional[List] = None, updated_assets: Optional[List] = None
+):
+    """Applies the provided changes to a parent asset graph. We do this by getting the list of
+    assets from the parent Definitions object and adding in the new_assets and replacing an
+    updated_assets. Then we can make a new Definitions object and make an AssetGraph.
+    """
+    parent_asset_graph = parent_deployment_defs_by_name[parent_graph_name].get_asset_graph()
+    parent_assets_by_key = {
+        asset.key: asset
+        for asset in list(parent_asset_graph.assets) + list(parent_asset_graph.source_assets)
+    }
+    new_assets = new_assets or []
+    if updated_assets:
+        for asset in updated_assets:
+            if parent_assets_by_key.get(asset.key) is not None:
+                del parent_assets_by_key[asset.key]
+            else:
+                assert False, "Asset included in updated_assets must exist as either an asset or a source asset in parent deployment"
+    else:
+        updated_assets = []
+    return AssetGraph.from_assets(
+        all_assets=new_assets + updated_assets + list(parent_assets_by_key.values())
+    )
```

Comment:
my intuition was that we would want to just have N files like [1] with modifications and compare them to each other instead of doing mutative calls like this. I think my argument would be that its more likely this test will not need future updating if it instantiates things like users do versus making mutations to internal data structures. 

That said, I can see how this mutative set up would make it much easier to rattle through different test permutations without adding a bunch of files / code so will defer to you to make the call

Reply:
yeah i think i might need to split out into files anyway since i think i need to lock down the typing of the branch and parent graphs to `ExternalAssetGraph` not just `AssetGraph`. and so i probably need to go through the whole making a workspace stuff for the branch graph now too vs just making an InternalAssetGraph from a list of assets.

Reply:
ended up moving this into a bunch of files containing the various changes. Interested in your thoughts on the new structure. I'm not 100% happy with it, but it gives a lot more flexibility to test more realistic scenarios